### PR TITLE
improve findSimilarClasses with awareness of class channel name format

### LINKED ISSUE #835: improve findSimilarClasses with awareness of class channel name format
public findSimilarClasses(className: string) {

ideally we try to tokenise the className so that if the input guess (also rename the param from className to maybeClassName or classNameGuess , pls ) has part of a class channel name then it filters down further

so inputs like 
- input: cap 
- output: cap1234, cap1567, etc 

whereas:
- input: 3502
- output: cop3502, cap3502, eng3502 

and like
- input: szumlanski (rip) 
- output: cop3502_szumlanski , cop3503_szumlanski

### PR DESCRIPTION
**Pull request by msabur**
Improve findSimilarClasses to support narrowing down class channels based on prefix (e.g. cop), code (e.g. 3223), professor name, the entire name, or a prefix of the entire name (e.g. cop3223). Keeps Levenshtein similarity as a fallback.

Closes #835

### CODE REVIEW

File: src/services/class.service.ts [hunk 1]
```diff
@@ -323,6 +323,20 @@ export class ClassService {
     return `You have successfully been removed from the ${categoryType} category.`;
   }
 
+  private _tokenizeClassChannelName(channelName: string) {
+    const tokenizingRegex = /^([a-z]{3})([0-9]{4}h?)_([a-z]+)$/;
```
Comment (tanndlin) [inline src/services/class.service.ts:327]: This [already exists](https://github.com/cs-discord-at-ucf/lion/blob/5b13a8c7ba254e646bf8659c0a011820153ead34/src/app/commands/delclasschans.command.ts#L7) somewhere. Could you remove the one in delclasschans, and put yours in `constants.ts` since yours is slightly better

**Review by tanndlin — APPROVED** [review]
pog

File: src/services/class.service.ts [hunk 2]
```diff
@@ -335,14 +349,34 @@ export class ClassService {
...
     const classes: string[] = Array.from(this.getClasses(types.ClassType.ALL).keys());
 
-    // Returns 10 most likely classes, if the caller wants less it can manage it.
-    return classes
-      .sort((a: string, b: string) => levenshtein(className, a) - levenshtein(className, b))
-      .splice(0, 10);
+    let results: string[] = [];
+
+    // Handle guesses of these forms: 'cop', '4934', 'adam', 'cop4934_adam'
+    for (const c of classes) {
```
Comment (c650) [inline src/services/class.service.ts:361]:
```suggestion
    for (const className of classes) {
```

**Review by c650 — APPROVED** [review]
this lgtm

it would be fantastic if you could add a few unit tests that have an input and list of classes, or if you otherwise show how you tested this locally

**Comment by msabur** [comment]
> this lgtm
> 
> it would be fantastic if you could add a few unit tests that have an input and list of classes, or if you otherwise show how you tested this 

Added some unit tests. 
To support ClassService I stubbed GuildService including guild.channels.cache. I'm not that experienced in writing tests, so I would appreciate any feedback on my approach or suggestions for improvement.

Here are also a few screenshots from my local testing using the server's template at https://discord.com/template/y79RbsjxvZfN :

<details>
  <summary>screenshots</summary>
  
  
Screenshot from 2024-09-27 01-59-29
Screenshot from 2024-09-27 01-59-45
Screenshot from 2024-09-27 01-59-53
Screenshot from 2024-09-27 02-00-29
Screenshot from 2024-09-27 02-00-48

  
</details>

File: src/__tests__/findSimilarClasses.test.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import { ClassService } from '../services/class.service';
+import { LoggerService } from '../services/logger.service';
+import { GuildService } from '../services/guild.service';
+import { MessageService } from '../services/message.service';
+
+const sampleClasses = [
+  'cot4593_umar',
+  'cot4593_zaid',
+  'cot2444_bradford',
+  'cot4593h_silicatony',
```
Comment (c650) [inline src/__tests__/findSimilarClasses.test.ts:11]: ah the `h` is a good edge case

File: src/__tests__/findSimilarClasses.test.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  'cap4453_norman',
+  'cap4453_khan',
+  'cap3223_spectre',
+  'cap3223_umar'
+];
+
+const root_id = 3333, category_id = 9999;
```
Comment (c650) [inline src/__tests__/findSimilarClasses.test.ts:18]: nit - camelCase ?

**Review by c650 — APPROVED** [review]

**Review by c650 — APPROVED** [review]
